### ISSUE TITLE
No way to export a private key (to refund a submarine swap)

### ISSUE THREAD

**Issue by interfect** [btcontract/lnwallet#132]
Since the wallet has no built-in on-chain/off-chain swap feature, I attempted to use the [Submarine Swaps](https://submarineswaps.org/) service to get a lightning payment sent to me in exchange for an on-chain payment. Not wanting to have to deal with a paper wallet on my phone, I plugged in my receive address from BLW as the refund address.

The swap didn't go through, presumably because I was on a moving bus and/or had my phone in my pocket with the screen off when my transaction finally confirmed and my counterparty tried to pay my invoice. So, I had to execute the refund.

The Submarine Swaps page has a simple way to do that. But it's a wrapper around an interaction with an on-chain smart contract *on behalf of the address you entered as the refund address*. So in order to actually *execute* the refund, you either need to get ahold of your raw private key for that address, or you need to have functionality to interact with this kind of contract written into your wallet. But having done that, you can *send* the refund anywhere.

Because I enjoy having my funds stolen, I decided to give this extremely trustworthy default Bootstrap website the private key for the address, to let it send the refund for me. But (and here's the actual bug report/feature request) **there's no way to get the private key for a single address out of BLW**. I had to import my seed phrase into a more versatile program (and guess the BIP44 derivation path that BLW uses; it's [BIP84](https://github.com/bitcoin/bips/blob/master/bip-0084.mediawiki)) to get ahold of the right key, which seems like a greater security risk than just letting me get at the key in BLW. Since BLW also can't export the extended public key for the account, exporting a private key and giving it away will grant access only to funds in that particular address, and not the whole wallet.

This is relevant to #62, because I was trying to work around it by doing a swap, and also to any issue about more general address or coin control features.

**Comment by interfect** [comment]
FWIW, Submarine Swaps doesn't seem to work any better with the phone plugged in and sitting on a table on wifi. The site tries to send to me over Lightning, and I can see my channel in the BLW flashing some numbers for "value in-flight", but apparently 100 consecutive attempts to pay my invoice fail.

**Comment by btcontract** [comment]
I see. My hope is splicing will be added to spec sooner rather than later to enable users to send local on-chain transactions to (and from) existing channels rather than deal with private keys and third-party swaps.

> The swap didn't go through, presumably because I was on a moving bus and/or had my phone in my pocket with the screen off when my transaction finally confirmed and my counterparty tried to pay my invoice. So, I had to execute the refund.

This is unlikely unless there were connection issues. Once incoming LN payment is requested BLW shows a notification which in turn keeps an app running until a payment arrives, even if phone screen is turned off.

> The site tries to send to me over Lightning, and I can see my channel in the BLW flashing some numbers for "value in-flight", but apparently 100 consecutive attempts to pay my invoice fail.

This is an interesting moment. "value in-flight" refers to both incoming and outgoing LN payments and represents amounts which are present in channel but not yet settled. It may show an incoming payment amount but for a very little period of time because BLW fulfills any incoming payment immediately.

If you see non-empty "value in-flight" for a long time it most likely represents an outgoing payment which you have sent to someone and it got stuck along a payment route. 

If this is indeed related to incoming payment a service is trying to send to you then it's a serious bug, BLW should not delay such payments. Can you make sure there are no stuck outgoing payments and this "value in-flight" is definitely about an incoming payment?

**Comment by interfect** [comment]
The value in flight only appears momentarily before going back to "nothing" when the payment fails. The diagram at the top has the right hand black area briefly expand, and then go back to its original size.

I now think the underlying problem might have been the counterparty on the channel I was trying to receive with. If they (or someone between Submarine Swaps and them) showed up as a valid route but didn't actually route the payment, this is the behavior I would expect, right?

February 28, 2019 1:30 AM, "Anton Kumaigorodski" <notifications@github.com (mailto:notifications@github.com?to=%22Anton%20Kumaigorodski%22%20<notifications@github.com>)> wrote:
	I see. My hope is splicing will be added to spec sooner rather than later to enable users to send local on-chain transactions to (and from) existing channels rather than deal with private keys and third-party swaps. 

	The swap didn't go through, presumably because I was on a moving bus and/or had my phone in my pocket with the screen off when my transaction finally confirmed and my counterparty tried to pay my invoice. So, I had to execute the refund. 

	This is unlikely unless there were connection issues. Once incoming LN payment is requested BLW shows a notification which in turn keeps an app running until a payment arrives, even if phone screen is turned off. 

	The site tries to send to me over Lightning, and I can see my channel in the BLW flashing some numbers for "value in-flight", but apparently 100 consecutive attempts to pay my invoice fail. 

	This is an interesting moment. "value in-flight" refers to both incoming and outgoing LN payments and represents amounts which are present in channel but not yet settled. It may show an incoming payment amount but for a very little period of time because BLW fulfills any incoming payment immediately. 

	If you see non-empty "value in-flight" for a long time it most likely represents an outgoing payment which you have sent to someone and it got stuck along a payment route. 

	If this is indeed related to incoming payment a service is trying to send to you then it's a serious bug, BLW should not delay such payments. Can you make sure there are no stuck outgoing payments and this "value in-flight" is definitely about an incoming payment?

**Comment by btcontract** [comment]
> they (or someone between Submarine Swaps and them) showed up as a valid route but didn't actually route the payment, this is the behavior I would expect, right?

No, the fact that value in-flight changes (as well as diagram) means a direct peer sends a payment in but BLW fails it right away.

Two possible reasons for this:
1. BLW does not have a corresponding payment preimage in a database. This is very unlikely as preimage is recorded to db before payment request is generated and then stored indefinitely.

2. Peer sends a payment, but not the one you have requested, perhaps swap service has mixed things up and tries to send you some unrelated payment (which is also very strange as payment details should be taken from payment request only, very little room for error here).

**Comment by interfect** [comment]
Hm. Unless there's a way to pull the logs out of the app, this underlying problem may have to remain a mystery. I've been in contact with my channel counterparty (Joltfun) and they said they route payments normally, and what you say confirms that.

February 28, 2019 8:50 PM, "Anton Kumaigorodski" <notifications@github.com (mailto:notifications@github.com?to=%22Anton%20Kumaigorodski%22%20<notifications@github.com>)> wrote:
	they (or someone between Submarine Swaps and them) showed up as a valid route but didn't actually route the payment, this is the behavior I would expect, right? 

	No, the fact that value in-flight changes (as well as diagram) means a direct peer sends a payment in but BLW fails it right away. 

	Two possible reasons for this: 
	* 

	BLW does not have a corresponding payment preimage in a database. This is very unlikely as preimage is recorded to db before payment request is generated and then stored indefinitely. 
	* 

	Peer sends a payment, but not the one you have requested, perhaps swap service has mixed things up and tries to send you some unrelated payment (which is also very strange as payment details should be taken from payment request only, very little room for error here).

**Comment by btcontract** [comment]
> Unless there's a way to pull the logs out of the app

Sally there are no logs.

> I've been in contact with my channel counterparty (Joltfun) and they said they route payments normally.

Your direct peer has nothing to do with this, unless peer is also a payer. Would it be possible to contact a swap service owner and provide them with some small (like 1 sat) invoice, and then see if BLW accepts it? I'd appreciate a lot of you could pull this off.
